core/common/src/main/java/alluxio/collections/UnmodifiableArrayList.java

```diff
@@ -0,0 +1,311 @@
+  }
+
+  @Override
+  public void clear() {
+    throw new UnsupportedOperationException(
+        "modification is not supported in UnmodifiableArrayList");
+  }
+
+  @Override
+  public T get(int index) {
+    return mElements[index];
+  }
+
+  @Override
+  public T set(int index, T element) {
+    throw new UnsupportedOperationException(
+        "modification is not supported in UnmodifiableArrayList");
+  }
+
+  @Override
+  public int indexOf(Object o) {
+    if (o == null) {
+      for (int i = 0; i < mElements.length; i++) {
+        if (mElements[i] == null) {
+          return i;
+        }
+      }
+    } else {
+      for (int i = 0; i < mElements.length; i++) {
+        if (o.equals(mElements[i])) {
+          return i;
+        }
+      }
+    }
+    return -1;
+  }
+
+  @Override
+  public int lastIndexOf(Object o) {
+    if (o == null) {
+      for (int i = mElements.length - 1; i >= 0; i--) {
+        if (mElements[i] == null) {
+          return i;
+        }
+      }
+    } else {
+      for (int i = mElements.length - 1; i >= 0; i--) {
+        if (o.equals(mElements[i])) {
+          return i;
+        }
+      }
+    }
+    return -1;
+  }
+
+  @Override
+  public ListIterator<T> listIterator() {
+    return new ListItr(0);
+  }
+
+  @Override
+  public ListIterator<T> listIterator(int index) {
+    if (index < 0 || index > mElements.length) {
+      throw new IndexOutOfBoundsException("Index: " + index);
+    }
+    return new ListItr(index);
+  }
+
+  @Override
+  public List<T> subList(int fromIndex, int toIndex) {
+    throw new UnsupportedOperationException(
+        "subList is not supported in UnmodifiableArrayList, use the underlying array instead");
+  }
+
+  private class Itr implements Iterator<T> {
+    int mCursor;       // index of next element to return
+    int mLastRet = -1; // index of last element returned; -1 if no such
+
+    Itr() {}
+
+    @Override
+    public boolean hasNext() {
+      return mCursor != mElements.length;
+    }
+
+    @Override
+    public T next() {
+      int i = mCursor;
+      if (i >= mElements.length) {
+        throw new NoSuchElementException();
+      }
+      mCursor = i + 1;
+      return mElements[mLastRet = i];
+    }
+
+    @Override
+    public void remove() {
+      throw new UnsupportedOperationException(
+          "modification is not supported in UnmodifiableArrayList");
+    }
+
+    @Override
+    public void forEachRemaining(Consumer<? super T> consumer) {
+      Objects.requireNonNull(consumer);
+      final int size = mElements.length;
+      int i = mCursor;
+      if (i >= size) {
+        return;
+      }
+      while (i != size) {
+        consumer.accept(mElements[i++]);
+      }
+      // update once at end of iteration to reduce heap write traffic
+      mCursor = i;
+      mLastRet = i - 1;
+    }
+  }
+
+  private class ListItr extends Itr implements ListIterator<T> {
+    ListItr(int index) {
+      super();
+      mCursor = index;
+    }
```

Comment:
These two classes are actually more or less copied from `ArrayList`, therefore you see this flavor maybe a bit different from the Alluxio code base. It's just the quickest way...